src/states/establishing_node.rs

```diff
@@ -0,0 +1,463 @@
+            self.dispatch_routing_message(routing_msg, outbox)
+        } else {
+            Ok(Transition::Stay)
+        }
+    }
+}
+
+impl Bootstrapped for EstablishingNode {
+    fn ack_mgr(&self) -> &AckManager {
+        &self.ack_mgr
+    }
+
+    fn ack_mgr_mut(&mut self) -> &mut AckManager {
+        &mut self.ack_mgr
+    }
+
+    fn routing_msg_filter(&mut self) -> &mut RoutingMessageFilter {
+        &mut self.routing_msg_filter
+    }
+
+    fn timer(&mut self) -> &mut Timer {
+        &mut self.timer
+    }
+
+    fn send_routing_message_via_route(
+        &mut self,
+        routing_msg: RoutingMessage,
+        route: u8,
+        expires_at: Option<Instant>,
+    ) -> Result<(), RoutingError> {
+        self.send_routing_message_via_proxy(routing_msg, route, expires_at)
+    }
+}
+
+impl Relocated for EstablishingNode {
+    fn peer_mgr(&self) -> &PeerManager {
+        &self.peer_mgr
+    }
+
+    fn peer_mgr_mut(&mut self) -> &mut PeerManager {
+        &mut self.peer_mgr
+    }
+
+    fn process_connection(&mut self, pub_id: PublicId, outbox: &mut EventBox) {
+        if self.chain.is_peer_valid(&pub_id) {
+            self.add_to_routing_table(&pub_id, outbox);
+        }
+    }
+
+    fn is_peer_valid(&self, _pub_id: &PublicId) -> bool {
+        true
+    }
+
+    fn add_to_notified_nodes(&mut self, pub_id: PublicId) -> bool {
+        self.notified_nodes.insert(pub_id)
+    }
+
+    fn remove_from_notified_nodes(&mut self, pub_id: &PublicId) -> bool {
+        self.notified_nodes.remove(pub_id)
+    }
+
+    fn add_to_routing_table_success(&mut self, _: &PublicId) {}
+
+    fn add_to_routing_table_failure(&mut self, pub_id: &PublicId) {
+        self.disconnect_peer(pub_id)
+    }
+}
+
+impl NotEstablished for EstablishingNode {
+    const SEND_ACK: bool = true;
+
+    fn get_proxy_public_id(&self, proxy_name: &XorName) -> Result<&PublicId, RoutingError> {
+        proxied::find_proxy_public_id(self, &self.peer_mgr, proxy_name)
+    }
+}
+
+impl RelocatedNotEstablished for EstablishingNode {
+    fn our_prefix(&self) -> &Prefix<XorName> {
+        self.chain.our_prefix()
+    }
+
+    fn push_message_to_backlog(&mut self, msg: RoutingMessage) {
+        self.msg_backlog.push(msg)
+    }
+}
+
+impl Approved for EstablishingNode {
+    fn parsec_poll_one(&mut self) -> Option<Block> {
+        self.parsec.poll()
+    }
+
+    fn chain_mut(&mut self) -> &mut Chain {
+        &mut self.chain
+    }
+
+    fn handle_online_event(
+        &mut self,
+        new_pub_id: PublicId,
+        _: Authority<XorName>,
+        _: &mut EventBox,
+    ) -> Result<(), RoutingError> {
+        let _ = self.chain.add_member(new_pub_id)?;
+        Ok(())
+    }
+
+    fn handle_offline_event(
+        &mut self,
+        pub_id: PublicId,
+        _: &mut EventBox,
+    ) -> Result<(), RoutingError> {
+        let _ = self.chain.remove_member(pub_id)?;
+        Ok(())
+    }
+
+    fn handle_section_info_event(
```

Comment:
from node.rs
We use to do quite a few things before, it is not clear to me why they were not necessary, or where they moved.

```rust
    fn handle_section_info_event(
        &mut self,
        sec_info: SectionInfo,
        old_pfx: Prefix<XorName>,
        outbox: &mut EventBox,
    ) -> Result<Transition, RoutingError> {
        if !self.peer_mgr.is_established() && self.chain.is_member() {
            // We have just found a `SectionInfo` block that contains us. Now we can supply our
            // votes for all latest neighbour infos that have accumulated so far.
            let ni_events = self
                .chain
                .neighbour_infos()
                .map(|ni| ni.clone().into_network_event())
                .collect_vec();

            ni_events.into_iter().for_each(|ni_event| {
                self.vote_for_event(ni_event);
            });

            self.node_established(outbox);
        }

        if sec_info.prefix().is_extension_of(&old_pfx) {
            self.finalise_prefix_change()?;
            outbox.send_event(Event::SectionSplit(*sec_info.prefix()));
            self.send_neighbour_infos();
        } else if old_pfx.is_extension_of(sec_info.prefix()) {
            self.finalise_prefix_change()?;
            outbox.send_event(Event::SectionMerge(*sec_info.prefix()));
        }

        let our_name = *self.full_id.public_id().name();
        let self_sec_update = sec_info.prefix().matches(&our_name);

        if self.chain.is_member() {
            self.update_peer_states(outbox);

            if self_sec_update {
                self.send_neighbour_infos();
            } else {
                // Vote for neighbour update if we haven't done so already.
                // vote_for_event is expected to only generate a new vote if required.
                self.vote_for_event(sec_info.into_network_event());
            }
        }

        let _ = self.merge_if_necessary();

        Ok(Transition::Stay)
    }
```

Reply:
Could you comment on that?

Reply:
never mind, saw the comment bellow where it moved.